### PR TITLE
Recognize MacCatalyst as a superset of iOS

### PR DESCRIPTION
**Pull request by jeffhandley**
1. `OperatingSystem.IsIOS()` now returns `true` when running on MacCatalyst
2. `OperatingSystem.IsOSPlatform("ios")` now returns true when running on MacCatalyst
3. `[SupportedOSPlatformGuard]` attributes are applied to `OperatingSystem.IsIOS()` and `OperatingSystem.IsIOSVersionAtLeast()` to indicate those methods guard for either "ios" or "maccatalyst"

This contributes to https://github.com/dotnet/runtime/issues/53084#issuecomment-877750424.

### CODE REVIEW

**Review by akoeplinger — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 1]
```diff
@@ -4959,13 +4959,18 @@ public sealed partial class OperatingSystem : System.ICloneable, System.Runtime.
         public override string ToString() { throw null; }
         public static bool IsOSPlatform(string platform) { throw null; }
         public static bool IsOSPlatformVersionAtLeast(string platform, int major, int minor = 0, int build = 0, int revision = 0) { throw null; }
+
```
[the remark this answers is not in the document]
Comment (jeffhandley) [reply src/libraries/System.Runtime/ref/System.Runtime.cs]: Oops; thanks.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 2]
```diff
@@ -4959,13 +4959,18 @@ public sealed partial class OperatingSystem : System.ICloneable, System.Runtime.
...
         public static bool IsBrowser() { throw null; }
         public static bool IsLinux() { throw null; }
         public static bool IsFreeBSD() { throw null; }
         public static bool IsFreeBSDVersionAtLeast(int major, int minor = 0, int build = 0, int revision = 0) { throw null; }
         public static bool IsAndroid() { throw null; }
         public static bool IsAndroidVersionAtLeast(int major, int minor = 0, int build = 0, int revision = 0) { throw null; }
+        [System.Runtime.Versioning.SupportedOSPlatformGuardAttribute("ios")]
```
[the remark this answers is not in the document]
Comment (jeffhandley) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:4969]: I should have described it; sorry.

After talking with @buyaa-n, we concluded that we would have to have the fallback logic in place of inferring the single platform guard from the method name in the analyzer. That means all of the single platform annotations would be redundant with the conventional logic and would have no other practical value, but they would add to the assembly size (negligibly).

We can add them in for completeness if that's desired, but I figured we would do that when we circle back to model the macOS/OSX alias through the guard attributes too.

File: src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs [hunk 3]
```diff
@@ -172,18 +177,22 @@ public static bool IsAndroidVersionAtLeast(int major, int minor = 0, int build =
             => IsAndroid() && IsOSVersionAtLeast(major, minor, build, revision);
 
         /// <summary>
-        /// Indicates whether the current application is running on iOS.
+        /// Indicates whether the current application is running on iOS or MacCatalyst.
         /// </summary>
+        [SupportedOSPlatformGuard("ios")]
```
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:182]: I would prefer to remove it as we are discussed to not make the analyzer fully dependent on Guard attributes (at least not within 6.0)
```suggestion
```

File: src/libraries/System.Runtime/ref/System.Runtime.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:4969]: Yep, and for the same reason, we don't need `[SupportedOSPlatformGuard("ios")]`

File: src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs (hunk 3 above)
Comment (jeffhandley) [reply src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:182]: Oh, and literally only apply the guard attributes for _additional_ platforms? Makes sense.

File: src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs [hunk 4]
```diff
@@ -101,6 +102,10 @@ public static bool IsOSPlatform(string platform)
             return platform.Equals("WINDOWS", StringComparison.OrdinalIgnoreCase);
 #elif TARGET_OSX
             return platform.Equals("OSX", StringComparison.OrdinalIgnoreCase) || platform.Equals("MACOS", StringComparison.OrdinalIgnoreCase);
+#elif TARGET_MACCATALYST
+            return platform.Equals("MACCATALYST", StringComparison.OrdinalIgnoreCase) || platform.Equals("IOS", StringComparison.OrdinalIgnoreCase);
+#elif TARGET_IOS
+            return platform.Equals("IOS", StringComparison.OrdinalIgnoreCase);
```
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:108]: I thought we want this:
```suggestion
#elif TARGET_MACCATALYST
            return platform.Equals("MACCATALYST", StringComparison.OrdinalIgnoreCase);
#elif TARGET_IOS
            return platform.Equals("IOS", StringComparison.OrdinalIgnoreCase) || platform.Equals("MACCATALYST", StringComparison.OrdinalIgnoreCase);
```

Comment (jeffhandley) [reply src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:108]: It's correct as it is, allowing "is this IOS" to return `true` on both IOS and MacCatalyst, but "is this MacCatalyst" will only return `true` when on MacCatalyst.

- When running on MacCatalyst
    - `IsOSPlatform("maccatalyst") => true`
    - `IsOSPlatform("ios") => true`
- When running on IOS
    - `IsOSPlatform("ios") => true`
    - `IsOSPlatform("maccatalyst") => false`

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs:108]: Never mind I confused 😝

**Review by buyaa-n — APPROVED** [review]
LGTM

**Comment by jeffhandley** [comment]
The test failures are Quic tests timing out and they are unrelated.

**Pull request merged**